rewriter/Concern.cc

```diff
@@ -0,0 +1,152 @@
+    ast::TreePtr classMethodsNode;
+    for (auto &stat : klass->rhs) {
+        if (auto *send = ast::cast_tree<ast::Send>(stat)) {
+            if (send->fun == core::Names::classMethodsTemp()) { // class_methods do ... end
+                if (send->block == nullptr) {
+                    continue;
+                }
+                if (!send->recv.isSelfReference()) {
+                    continue;
+                }
+
+                if (classMethodsNode) {
+                    // ClassMethods module already exists. Let's add the block as one of its members
+                    storeBlockInExistingModule(send, classMethodsNode);
+                } else {
+                    // We don't have a ClassMethods module yet. Let's create it and add the send definition to its
+                    // members
+                    auto loc = send->loc;
+                    generateSendForExtendHelpers(loc, stats);
+                    classMethodsNode = generateClassMethodsWithMethodBody(ctx, loc, send);
+                }
+                continue;
+            }
+        } else if (auto *mod = ast::cast_tree<ast::ClassDef>(stat)) {
+            auto name = ast::cast_tree<ast::UnresolvedConstantLit>(mod->name);
+            if (name && name->cnst == core::Names::Constants::ClassMethods()) {
+                if (classMethodsNode) {
+                    // ClassMethods module already exists. Let's add mod->rhs into existing module
+                    storeRhsInExistingModule(mod, classMethodsNode);
+                } else {
+                    // We don't have a ClassMethods module yet. Let's use the defined module (mod)
+                    auto loc = mod->loc;
+                    generateSendForExtendHelpers(loc, stats);
+                    // Save the module
+                    classMethodsNode = ast::MK::Module(loc, loc, std::move(mod->name), {}, std::move(mod->rhs));
+                }
+                continue;
+            }
+        }
+        // Everything else that we are not interested in should be copied as is
+        stats.emplace_back(std::move(stat));
+    }
+
+    // Generate a Send { mixes_in_class_methods(ClassMethods) }
+    if (classMethodsNode) {
+        generateSendForMixesInClassMethods(klass->loc, stats);
+    }
+
+    storeNewClassDef(klass, &stats, classMethodsNode);
+}
+
+bool Concern::doesExtendConcern(core::MutableContext ctx, ast::Send *send) {
+    if (send->fun != core::Names::extend()) {
+        return false;
+    }
+
+    auto &args = send->args;
+    if (args.empty()) {
+        return false;
+    }
+    auto *firstArg = ast::cast_tree<ast::UnresolvedConstantLit>(args.at(0));
+    if (!firstArg) {
+        return false;
+    }
+    auto *scope = ast::cast_tree<ast::UnresolvedConstantLit>(firstArg->scope);
+    if (!scope) {
+        return false;
+    }
+    if (firstArg->cnst != core::Names::Constants::Concern() || scope->cnst != core::Names::Constants::ActiveSupport()) {
+        return false;
+    }
```

Comment:
What happens if I write this:

```ruby
extend DefinitelyNot::ActiveSupport::Concern
```

Does this rewriter pass still fire?

Reply:
It did fire, thank you. I opted for a similar approach as before to fix this. (`scope` is now renamed to `firstArgScope`)

```
auto *outerScope = ast::cast_tree<ast::UnresolvedConstantLit>(firstArgScope->scope);
if (outerScope != nullptr) {
    return;
}
```